docs/Architecture.md

```diff
@@ -0,0 +1,233 @@
+
+- Service HTTP port, for setting the default HTTP components port to `50000` with `PATRON_HTTP_DEFAULT_PORT`
+- Service HTTP read and write timeout, for changing the default HTTP read and write timeout to any value, use `PATRON_HTTP_READ_TIMEOUT`, `PATRON_HTTP_WRITE_TIMEOUT` respectively. For acceptable values check [here](https://golang.org/pkg/time/#ParseDuration).
+- Log level, for setting zerolog with `INFO` log level with `PATRON_LOG_LEVEL`
+- Tracing, for setting up jaeger tracing with
+  - agent host `0.0.0.0` with `PATRON_JAEGER_AGENT_HOST`
+  - agent port `6831` with `PATRON_JAEGER_AGENT_PORT`
+  - sampler type `probabilistic`with `PATRON_JAEGER_SAMPLER_TYPE`
+  - sampler param `0.0` with `PATRON_JAEGER_SAMPLER_PARAM`, which means that traces are not initiated here.
+
+### Component
+
+A `Component` is an interface that exposes the following API:
+
+```go
+type Component interface {
+  Run(ctx context.Context) error  
+}
+```
+
+The above API gives the `Service` the ability to start and gracefully shutdown a `component` via context cancellation. Furthermore, the component describes itself by implementing the `Info` method and thus giving the service the ability to report the information of all components. The framework divides the components in 2 categories:
+
+- synchronous, which are components that follow the request/response pattern and
+- asynchronous, which consume messages from a source but don't respond anything back
+
+The following component implementations are available:
+
+- HTTP (sync)
+- gRPC
+- RabbitMQ consumer (async)
+- Kafka consumer (async)
+- AWS SQS (async)
+
+Adding to the above list is as easy as implementing a `Component` and a `Processor` for that component.
+
+## Processors
+
+### Synchronous
+
+The implementation of the processor is responsible to create a `Request` by providing everything that is needed (Headers, Fields, decoder, raw io.Reader) pass it to the implementation by invoking the `Process` method and handle the `Response` or the `error` returned by the processor.
+
+The sync package contains only a function definition along with the models needed:
+
+```go
+type ProcessorFunc func(context.Context, *Request) (*Response, error)
+```
+
+The `Request` model contains the following properties (which are provided when calling the "constructor" `NewRequest`)
+
+- Fields, which may contain any fields associated with the request
+- Raw, the raw request data (if any) in the form of a `io.Reader`
+- Headers, the request headers in the form of `map[string]string`
+- decode, which is a function of type `encoding.Decode` that decodes the raw reader
+
+An exported function exists for decoding the raw io.Reader in the form of
+
+```go
+Decode(v interface{}) error
+```
+
+The `Response` model contains the following properties (which are provided when calling the "constructor" `NewResponse`)
+
+- Payload, which may hold a struct of type `interface{}`
+
+### Middlewares per Route
```

Comment:
i missed how we ended up in the routes, i would expect an HTTP section somewhere above. Or am i missing something ?

Reply:
documentation is still in WIP. The description contains the unfinished work.